File: common/src/test/java/DahRealPanda/plugins/colonyweb/auth/SessionCookieTest.java
package DahRealPanda.plugins.colonyweb.auth;

import DahRealPanda.plugins.colonyweb.testsupport.FakeHttpExchange;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Parsing the session cookie out of a request, and the flags set when issuing one.
 *
 * <p>The cookie carries the only credential the dashboard has, so the flags on it are part of
 * the security model rather than cosmetic: {@code HttpOnly} keeps injected page scripts from
 * reading the token and {@code SameSite=Lax} keeps another site from riding the session.</p>
 */
class SessionCookieTest {

    @Nested
    @DisplayName("reading")
    class Reading {

        @Test
        @DisplayName("finds the token in a lone cookie")
        void singleCookie() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", SessionCookie.NAME + "=abc123");

            assertEquals(Optional.of("abc123"), SessionCookie.read(exchange));
        }

        @Test
        @DisplayName("finds the token among other cookies")
        void amongOthers() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", "theme=dark; " + SessionCookie.NAME + "=abc123; lang=en");

            assertEquals(Optional.of("abc123"), SessionCookie.read(exchange));
        }

        @Test
        @DisplayName("finds the token when the browser sends several Cookie headers")
        void multipleHeaders() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", "theme=dark")
                    .withHeader("Cookie", SessionCookie.NAME + "=abc123");

            assertEquals(Optional.of("abc123"), SessionCookie.read(exchange));
        }

        @Test
        @DisplayName("no Cookie header at all means no session")
        void noHeader() {
            assertTrue(SessionCookie.read(FakeHttpExchange.get("/api/colonies")).isEmpty());
        }

        @Test
        @DisplayName("an unrelated cookie is not mistaken for the session")
        void unrelatedCookie() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", "theme=dark");

            assertTrue(SessionCookie.read(exchange).isEmpty());
        }

        @Test
        @DisplayName("a cookie whose name merely ends with ours is not matched")
        void suffixNameIsNotMatched() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", "evil_" + SessionCookie.NAME + "=stolen");

            assertTrue(SessionCookie.read(exchange).isEmpty(),
                    "matching on a suffix would let any cookie impersonate the session");
        }

        @ParameterizedTest
        @DisplayName("an empty value is treated as no session")
        @ValueSource(strings = {"colonyweb_session=", "colonyweb_session=   ", "theme=dark; colonyweb_session="})
        void blankValue(String header) {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies").withHeader("Cookie", header);

            assertTrue(SessionCookie.read(exchange).isEmpty());
        }

        @Test
        @DisplayName("surrounding whitespace is tolerated")
        void trimsWhitespace() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", "  theme=dark ;   " + SessionCookie.NAME + "=abc123  ");

            assertEquals(Optional.of("abc123"), SessionCookie.read(exchange));
        }

        @Test
        @DisplayName("the first matching cookie wins")
        void firstMatchWins() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/api/colonies")
                    .withHeader("Cookie", SessionCookie.NAME + "=first; " + SessionCookie.NAME + "=second");

            assertEquals(Optional.of("first"), SessionCookie.read(exchange));
        }
    }

    @Nested
    @DisplayName("writing")
    class Writing {

        @Test
        @DisplayName("issuing a session sets the hardening flags")
        void setIsHardened() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/auth/login");

            SessionCookie.set(exchange, "abc123", 2_592_000L);

            String cookie = exchange.setCookies().get(0);
            assertTrue(cookie.startsWith(SessionCookie.NAME + "=abc123"), cookie);
            assertTrue(cookie.contains("Max-Age=2592000"), cookie);
            assertTrue(cookie.contains("HttpOnly"), "page scripts must not be able to read the token");
            assertTrue(cookie.contains("SameSite=Lax"), "another site must not be able to ride the session");
            assertTrue(cookie.contains("Path=/"), cookie);
        }

        @Test
        @DisplayName("clearing expires the cookie immediately")
        void clearExpires() {
            FakeHttpExchange exchange = FakeHttpExchange.get("/auth/logout");

            SessionCookie.clear(exchange);

            String cookie = exchange.setCookies().get(0);
            assertTrue(cookie.contains("Max-Age=0"), cookie);
            assertTrue(cookie.startsWith(SessionCookie.NAME + "="), cookie);
            assertTrue(cookie.contains("HttpOnly"), cookie);
        }

        @Test
        @DisplayName("a cookie this handler wrote is one the reader understands")
        void roundTrip() {
            FakeHttpExchange write = FakeHttpExchange.get("/auth/login");
            SessionCookie.set(write, "round-trip-token", 60);

            String issued = write.setCookies().get(0);
            String pair = issued.substring(0, issued.indexOf(';'));
            FakeHttpExchange read = FakeHttpExchange.get("/api/colonies").withHeader("Cookie", pair);

            assertEquals(Optional.of("round-trip-token"), SessionCookie.read(read));
        }
    }
}

File: common/src/test/kotlin/DahRealPanda/plugins/colonyweb/auth/SessionCookieTest.kt
package DahRealPanda.plugins.colonyweb.auth

import DahRealPanda.plugins.colonyweb.support.FakeHttpExchange
import com.sun.net.httpserver.Headers
import com.sun.net.httpserver.HttpExchange
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain

/**
 * The session cookie is the credential every dashboard request carries, so this covers both the
 * parsing (a browser sends its whole cookie jar, not just ours) and the attributes that keep the
 * token out of reach of page scripts and cross-site requests.
 */
class SessionCookieTest : DescribeSpec({

    fun exchangeWithCookies(vararg headers: String): HttpExchange =
        FakeHttpExchange.withCookieHeaders(*headers)

    fun responseHeadersOf(write: (HttpExchange) -> Unit): Headers {
        val exchange = FakeHttpExchange()
        write(exchange)
        return exchange.responseHeaders
    }

    describe("reading the token from a request") {
        it("finds the token when ours is the only cookie") {
            SessionCookie.read(exchangeWithCookies("colonyweb_session=abc123")) shouldBe "abc123"
        }

        it("finds the token among the browser's other cookies") {
            SessionCookie.read(
                exchangeWithCookies("theme=dark; colonyweb_session=abc123; locale=en")
            ) shouldBe "abc123"
        }

        it("finds the token when the browser splits cookies across headers") {
            SessionCookie.read(
                exchangeWithCookies("theme=dark", "colonyweb_session=abc123")
            ) shouldBe "abc123"
        }

        it("returns null when the request carries no cookies at all") {
            SessionCookie.read(exchangeWithCookies()) shouldBe null
        }

        it("returns null when none of the cookies is ours") {
            SessionCookie.read(exchangeWithCookies("theme=dark; locale=en")) shouldBe null
        }

        // A cleared cookie arrives as an empty value; treating it as a token would send an
        // obviously-invalid credential to the auth service on every request.
        it("treats an empty value as no session") {
            SessionCookie.read(exchangeWithCookies("colonyweb_session=")) shouldBe null
        }

        it("does not mistake a cookie whose name merely starts the same") {
            SessionCookie.read(exchangeWithCookies("colonyweb_session_backup=abc123")) shouldBe null
        }

        it("does not mistake a cookie whose name ends with ours") {
            SessionCookie.read(exchangeWithCookies("other_colonyweb_session=abc123")) shouldBe null
        }

        it("keeps a token that itself contains an equals sign") {
            SessionCookie.read(exchangeWithCookies("colonyweb_session=abc=123")) shouldBe "abc=123"
        }
    }

    describe("writing the cookie") {
        it("sends the token back with a lifetime") {
            val headers = responseHeadersOf { SessionCookie.set(it, "abc123", 3600) }

            headers["Set-Cookie"]!! shouldHaveSize 1
            headers.getFirst("Set-Cookie") shouldContain "colonyweb_session=abc123"
            headers.getFirst("Set-Cookie") shouldContain "Max-Age=3600"
            headers.getFirst("Set-Cookie") shouldContain "Path=/"
        }

        // HttpOnly keeps page scripts — and anything injected into them — from reading the token;
        // SameSite=Lax keeps another site from riding the session with a cross-origin request.
        it("marks the cookie HttpOnly and SameSite=Lax") {
            val headers = responseHeadersOf { SessionCookie.set(it, "abc123", 3600) }

            headers.getFirst("Set-Cookie") shouldContain "HttpOnly"
            headers.getFirst("Set-Cookie") shouldContain "SameSite=Lax"
        }

        it("expires the cookie immediately on logout") {
            val headers = responseHeadersOf { SessionCookie.clear(it) }

            headers.getFirst("Set-Cookie") shouldContain "colonyweb_session=;"
            headers.getFirst("Set-Cookie") shouldContain "Max-Age=0"
        }

        it("keeps the security attributes when clearing, so the two headers match") {
            val headers = responseHeadersOf { SessionCookie.clear(it) }

            headers.getFirst("Set-Cookie") shouldContain "HttpOnly"
            headers.getFirst("Set-Cookie") shouldContain "SameSite=Lax"
            headers.getFirst("Set-Cookie") shouldContain "Path=/"
        }
    }
})
